source/common/network/cidr_range.h

```diff
@@ -0,0 +1,114 @@
+   */
+  CidrRange();
+
+  /**
+   * Copies an existing CidrRange.
+   */
+  CidrRange(const CidrRange& other);
+
+  /**
+   * Overwrites this with other.
+   */
+  CidrRange& operator=(const CidrRange& other);
+
+  /**
+   * @return Ipv4 address data IFF length >= 0 and version() == IpVersion::v4, otherwise nullptr.
+   */
+  const Ipv4* ipv4() const;
+
+  /**
+   * @return Ipv6 address data IFF length >= 0 and version() == IpVersion::v6, otherwise nullptr.
+   */
+  const Ipv6* ipv6() const;
+
+  /**
+   * @return the number of bits of the address that are included in the mask.
+   * -1 if uninitialized or invalid, else in the range 0 to 32 for IPv4, and 0 to 128 for IPv6.
+   */
+  int length() const;
+
+  /**
+   * @return the version of IP address.
+   */
+  IpVersion version() const;
+
+  /**
+   * @return true if the address argument is in the range of this object.
+   * false if not, including if the range is uninitialized or if the argument is not of the
+   * same IpVersion.
+   */
+  bool isInRange(const InstancePtr& address) const;
+
+  /**
+   * @return a human readable string for the range.
+   *
+   * This string will be in the following format:
+   * For IPv4 ranges: "1.2.3.4/32" or "10.240.0.0/16"
+   * For IPv6 ranges: "1234:5678::f/128" or "1234:5678::/64"
+   */
+  std::string asString() const;
+
+  /**
+   * @return true if this instance is valid; address != nullptr && length is appropriate
+   * for the IP version (these are checked during construction, and reduced down to a
+   * check of the length).
+   */
+  bool isValid() const { return length_ >= 0; }
+
+  /**
+   * @return an CidrRange instance with the specified address and length, modified so that
+   * the only bits that might be non-zero are in the high-order length bits, and so that
+   * length is in the appropriate range (0 to 32 for IPv4, 0 to 128 for IPv6). If the
+   * the address or length is invalid, then the range will be invalid (i.e. length == -1).
+   */
+  static CidrRange construct(const InstancePtr& address, int length);
+  static CidrRange construct(const std::string& address, int length);
+
+  /**
+   * Constructs an CidrRange from a string with this format (same as returned
+   * by CidrRange::asString above):
+   *      <address>/<length>    e.g. "10.240.0.0/16" or "1234:5678::/64"
+   * @return an CidrRange instance with the specified address and length if parsed successfully,
+   * else with no address and a length of -1.
+   */
+  static CidrRange construct(const std::string& range);
+
+  /**
+   * Given an IP address and a length of high order bits to keep, returns an address
+   * where those high order bits are unmodified, and the remaining bits are all zero.
+   * length_io is reduced to be at most 32 for IPv4 address and at most 128 for IPv6
+   * addresses. If the address is invalid or the length is less than zero, then *length_io
+   * is set to -1 and nullptr is returned.
```

Comment:
I agree, but admit we have not been consistent. Ultimately it would be good to get Doxygen running as part of the build, then just have a docs bash and ultimately turn on doxygen warnings as errors.